src/solc_0.8/estate/EstateSaleWithAuth.sol

```diff
@@ -0,0 +1,290 @@
+            }
+        }
+
+        return computedHash == _merkleRoot;
+    }
+
+    function _handleFeeAndReferral(
+        address buyer,
+        uint256 priceInSand,
+        bytes memory referral
+    ) internal {
+        // send 5% fee to a specially configured instance of FeeDistributor.sol
+        uint256 remainingAmountInSand = _handleSandFee(buyer, priceInSand);
+
+        // calculate referral based on 95% of original priceInSand
+        handleReferralWithERC20(buyer, remainingAmountInSand, referral, _wallet, address(_sand));
+    }
+
+    function _handleSandFee(address buyer, uint256 priceInSand) internal returns (uint256) {
+        uint256 feeAmountInSand = (priceInSand * FEE) / 100;
+        require(_sand.transferFrom(buyer, address(_feeDistributor), feeAmountInSand), "FEE_TRANSFER_FAILED");
+        return priceInSand - feeAmountInSand;
+    }
+
+    uint256 internal constant GRID_SIZE = 408; // 408 is the size of the Land
+
+    IERC1155 internal immutable _asset;
+    LandToken internal immutable _land;
+    IERC20 internal immutable _sand;
+    address internal immutable _estate;
+    address internal immutable _feeDistributor;
+
+    address payable internal _wallet;
+    AuthValidator internal _authValidator;
+    uint256 internal immutable _expiryTime;
+    bytes32 internal immutable _merkleRoot;
+
+    uint256 private constant FEE = 5; // percentage of land sale price to be diverted to a specially configured instance of FeeDistributor, shown as an integer
+
+    // buyLandWithSand info indexes
+
+    uint256 private constant X_INDEX = 0;
+    uint256 private constant Y_INDEX = 1;
+    uint256 private constant SIZE_INDEX = 2;
+    uint256 private constant PRICE_INDEX = 3;
+
+    // need to use struct to avoid "Stack Too Deep" error. Since there are too many parameters.
+    struct Parameters {
+        IERC1155 asset;
+        LandToken landAddress;
+        IERC20 sandContractAddress;
+        address admin;
+        address estate;
+        address feeDistributor;
+        address payable initialWalletAddress;
+        AuthValidator authValidator;
+        uint256 expiryTime;
+        bytes32 merkleRoot;
+        address trustedForwarder;
+        address initialSigningWallet;
+        uint256 initialMaxCommissionRate;
+    }
+
+    constructor(Parameters memory p)
+        ERC2771Context(p.trustedForwarder)
+        ReferralValidator08(p.initialSigningWallet, p.initialMaxCommissionRate, p.admin)
+    {
+        _asset = p.asset;
+        _land = p.landAddress;
+        _sand = p.sandContractAddress;
+        // // ERC2771Handler.__ERC2771Handler_initialize(trustedForwarder);
+        _admin = p.admin;
+        _estate = p.estate;
+        _feeDistributor = p.feeDistributor;
+        _wallet = p.initialWalletAddress;
+        _authValidator = p.authValidator;
+        _expiryTime = p.expiryTime;
+        _merkleRoot = p.merkleRoot;
+    }
```

Comment:
Move properties and constructor at the beginning of the class

Reply:
fixed -> c21e258